### PR TITLE
[Microservices] Create service base class

### PR DESCRIPTION
**Pull request by alonmr**
Add a service and daemon base classes
The `Service` defines the interface and common code between microservices
The `Daemon` class initializes the service and shall configure it in the future

This is part 1 of https://iguazio.atlassian.net/browse/ML-8211

### CODE REVIEW

File: server/py/framework/service/__init__.py [hunk 1]
```diff
@@ -0,0 +1,192 @@
...
+            openapi_url=f"{self.SERVICE_PREFIX}/openapi.json",
+            docs_url=f"{self.SERVICE_PREFIX}/docs",
+            redoc_url=f"{self.SERVICE_PREFIX}/redoc",
+            default_response_class=fastapi.responses.ORJSONResponse,
+            lifespan=self.lifespan,
+        )
+
+    # https://fastapi.tiangolo.com/advanced/events/
+    @contextlib.asynccontextmanager
+    async def lifespan(self, app_: fastapi.FastAPI):
```
Comment (adamdelman) [inline server/py/framework/service/__init__.py:76]: Maybe add a return type annotation?

Comment (alonmr) [reply server/py/framework/service/__init__.py:76]: wdym? adding `-> None` ?

File: server/py/services/api/daemon.py [hunk 2]
```diff
@@ -0,0 +1,28 @@
...
+    def __init__(self, service_cls: framework.service.Service.__class__):
+        self._service = service_cls()
+
+    def initialize(self):
+        self._service.initialize()
+
+
+daemon = Daemon(service_cls=services.api.main.Service)
+daemon.initialize()
+app = daemon.app
```
Comment (liranbg) [inline server/py/services/api/daemon.py:32]:
```suggestion
if __name__ == "__main__":
  daemon = Daemon(service_cls=services.api.main.Service)
  daemon.initialize()
  app = daemon.app
```

Comment (alonmr) [reply server/py/services/api/daemon.py:32]: This doesn't work with our current uvircorn command. May be able to get it to work with the new fastapi CLI but it requires some investigation - can be done later IMO

**Review by liranbg — APPROVED** [review]

File: server/py/services/api/daemon.py [hunk 3]
```diff
@@ -0,0 +1,32 @@
...
+class Daemon(framework.service.Daemon):
+    def __init__(self, service_cls: framework.service.Service.__class__):
+        self._service = service_cls()
+
+    def initialize(self):
+        self._service.initialize()
+
+    @property
+    def service(self) -> services.api.main.Service:
+        return self._service
```
Comment (TomerShor) [inline server/py/services/api/daemon.py:27]: @alonmr Post-merge question: Why duplicate the same methods from the `framework.service.Daemon` class? I know it is an `ABC` but it seems redundant...

**Pull request merged**